File: java/src/main/java/aoc2021/day02/Navigation.java
package aoc2021.day02;

import library.Year;
import library.io.Input;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Navigation {

    private static List<Instruction> read() {
        Scanner sc = Input.openFile(Year._2021, "day02.txt");
        List<Instruction> data = new ArrayList<>();
        while (sc.hasNextLine()) {
            data.add(new Instruction(
                    Direction.valueOf(sc.next().toUpperCase()),
                    sc.nextInt()
            ));
        }
        return data;
    }

    private static int getFinalPosition(Position start, List<Instruction> instructions) {
        start.update(instructions);
        return start.getDepth() * start.getHorizontal();
    }

    public static void main(String[] args) {
        // load data
        List<Instruction> data = read();

        // Part 1
        System.out.println(getFinalPosition(new SimplePosition(), data));

        // Part 2
        System.out.println(getFinalPosition(new AimPosition(), data));
    }

}

File: kotlin/src/main/kotlin/aoc2021/day02/Navigation.kt
package aoc2021.day02

import library.Year
import library.readData

fun executeSimple(instructions: List<Instruction>) =
    instructions.fold(SimplePosition()) { pos: SimplePosition, instr: Instruction ->
        instr.command.simpleExecutor(pos, instr.steps)
    }

fun executeComplex(instructions: List<Instruction>) =
    instructions.fold(ComplexPosition()) { pos: ComplexPosition, instr: Instruction ->
        instr.command.complexExecutor(pos, instr.steps)
    }

fun main() {
    val instructions = readData(Year._2021, "day02.txt")
        .map { it.split(" ") }
        .map { Instruction(Command.valueOf(it[0].uppercase()), it[1].toInt()) }

    println(executeSimple(instructions).summary())
    println(executeComplex(instructions).summary())
}